File: storm-core/test/clj/backtype/storm/security/auth/DefaultHttpCredentialsPlugin_test.clj
(ns backtype.storm.security.auth.DefaultHttpCredentialsPlugin-test
  (:require [clojure.test :refer :all])
  (:import [javax.security.auth Subject]
           [javax.servlet.http HttpServletRequest]
           [backtype.storm.security.auth SingleUserPrincipal]
           [org.mockito Mockito]
           [backtype.storm.security.auth DefaultHttpCredentialsPlugin
                                         ReqContext SingleUserPrincipal]))

(deftest test-getUserName
  (let [handler (doto (DefaultHttpCredentialsPlugin.) (.prepare {}))]
    (testing "returns null when request is null"
      (is (nil? (.getUserName handler nil))))

    (testing "returns null when user principal is null"
      (let [req (Mockito/mock HttpServletRequest)]
        (is (nil? (.getUserName handler req)))))

    (testing "returns null when user is blank"
      (let [princ (SingleUserPrincipal. "")
            req (Mockito/mock HttpServletRequest)]
        (. (Mockito/when (. req getUserPrincipal))
           thenReturn princ)
        (is (nil? (.getUserName handler req)))))

    (testing "returns correct user from requests principal"
      (let [exp-name "Alice"
            princ (SingleUserPrincipal. exp-name)
            req (Mockito/mock HttpServletRequest)]
        (. (Mockito/when (. req getUserPrincipal))
           thenReturn princ)
        (is (.equals exp-name (.getUserName handler req)))))

    (testing "returns doAsUser from requests principal when Header has doAsUser param set"
      (let [exp-name "Alice"
            do-as-user-name "Bob"
            princ (SingleUserPrincipal. exp-name)
            req (Mockito/mock HttpServletRequest)
            _ (. (Mockito/when (. req getUserPrincipal))
              thenReturn princ)
            _ (. (Mockito/when (. req getHeader "doAsUser"))
              thenReturn do-as-user-name)
            context (.populateContext handler (ReqContext/context) req)]
        (is (= true (.isImpersonating context)))
        (is (.equals exp-name (.getName (.realPrincipal context))))
        (is (.equals do-as-user-name (.getName (.principal context))))))))

(deftest test-populate-req-context-on-null-user
  (let [req (Mockito/mock HttpServletRequest)
        handler (doto (DefaultHttpCredentialsPlugin.) (.prepare {}))
        subj (Subject. false (set [(SingleUserPrincipal. "test")]) (set []) (set []))
        context (ReqContext. subj)]
    (is (= 0 (-> handler (.populateContext context req) (.subject) (.getPrincipals) (.size))))))

File: storm-server/src/test/java/org/apache/storm/security/auth/DefaultHttpCredentialsPluginTest.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.  The ASF licenses this file to you under the Apache License, Version
 * 2.0 (the "License"); you may not use this file except in compliance with the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

package org.apache.storm.security.auth;

import java.security.Principal;
import java.util.HashMap;
import java.util.HashSet;
import javax.security.auth.Subject;
import javax.servlet.http.HttpServletRequest;
import org.apache.storm.shade.com.google.common.collect.ImmutableSet;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class DefaultHttpCredentialsPluginTest {

    @Test
    public void test_getUserName() {
        DefaultHttpCredentialsPlugin handler = new DefaultHttpCredentialsPlugin();
        handler.prepare(new HashMap<>());

        assertNull(handler.getUserName(null), "Should return null when request is null");

        assertNull(handler.getUserName(Mockito.mock(HttpServletRequest.class)), "Should return null when user principal is null");

        HttpServletRequest mockRequest = Mockito.mock(HttpServletRequest.class);
        Mockito.when(mockRequest.getUserPrincipal()).thenReturn(new SingleUserPrincipal(""));
        assertNull(handler.getUserName(mockRequest), "Should return null when user is blank");

        String expName = "Alice";
        mockRequest = Mockito.mock(HttpServletRequest.class);
        Mockito.when(mockRequest.getUserPrincipal()).thenReturn(new SingleUserPrincipal(expName));
        assertEquals(expName, handler.getUserName(mockRequest), "Should return correct user from requests principal");

        try {
            String doAsUserName = "Bob";
            mockRequest = Mockito.mock(HttpServletRequest.class);
            Mockito.when(mockRequest.getUserPrincipal()).thenReturn(new SingleUserPrincipal(expName));
            Mockito.when(mockRequest.getHeader("doAsUser")).thenReturn(doAsUserName);
            ReqContext context = handler.populateContext(ReqContext.context(), mockRequest);

            assertTrue(context.isImpersonating());
            assertEquals(expName, context.realPrincipal().getName());
            assertEquals(doAsUserName, context.principal().getName());
        } finally {
            ReqContext.reset();
        }
    }

    @Test
    public void test_populate_req_context_on_null_user() {
        try {
            DefaultHttpCredentialsPlugin handler = new DefaultHttpCredentialsPlugin();
            handler.prepare(new HashMap<>());
            Subject subject =
                new Subject(false, ImmutableSet.<Principal>of(new SingleUserPrincipal("test")), new HashSet<>(), new HashSet<>());
            ReqContext context = new ReqContext(subject);


            assertEquals(0, handler
                .populateContext(context, Mockito.mock(HttpServletRequest.class))
                .subject()
                .getPrincipals()
                .size()

            );
        } finally {
            ReqContext.reset();
        }

    }
}
